### PR TITLE
Fix infinite loop in event batch iterator

### PR DESCRIPTION
**Pull request by jonasrichard**
### Fix

When batch size is bigger than 1MB the event batch iterator went into an infinite loop. The iterator originally grouped all events by partitionKey and iterate over the groups and then inside the groups, creating EventBatch structures.

There were two errors in the original code: inside one group it iterated over the data in an infinite loop if the group was bigger than the max batch size. The other issue was that it iterated over a map which is not stable, so between iterations the iterator was unstable.

I reimplemented the iterator to keep using grouped events by partitionKey but I use array of grouped events (array of events) over which we can iterate in a stable way.

- [x] All tests passed
- [x] Add change to `changelog.md`

### Environment
- OS: Linux
- Go version: 1.13

### CODE REVIEW

**Review by adolsalamanca — APPROVED** [review]
Great work!!

**Comment by jonasrichard** [comment]
@catalinaperalta Hey, can you please take a look at this PR and see if we can merge it or we can add something what is missing.

File: batch.go [hunk 1]
```diff
@@ -19,8 +21,12 @@ type (
...
 	// EventBatchIterator provides an easy way to iterate over a slice of events to reliably create batches
 	EventBatchIterator struct {
-		Cursors            map[string]int
-		PartitionEventsMap map[string][]*Event
+		// Current chunk number
+		Cursor int
+		// Current position in the current chunk
+		PartitionCursor int
+		// Events grouped by the same partition key
+		PartitionEventChunks [][]*Event
```
Comment (catalinaperalta) [inline batch.go:29]: These are all breaking changes. Is it possible to use Cursors and PartitionEventsMap as they are now? Then add any additional fields to the struct as necessary?

Comment (jonasrichard) [reply batch.go]: I switched back to the original data structure. The only major change I made is to return with an error message if event is bigger than the max batch size and I step the iterator over (not having the infinite loop). Also I put the sub-partition fix what I found a previous PR.

**Comment by jonasrichard** [comment]
Hey, any feedback on this? I saw that it had been merged then later it was reopened. What can we do here now?

File: batch.go [hunk 2]
```diff
@@ -47,6 +49,8 @@ const (
 	KeyOfNoPartitionKey = "NoPartitionKey"
 )
 
+var ErrMessageIsTooBig = errors.New("message is too big")
```
Comment (jhendrixMSFT) [inline batch.go:52]: Can you please add a doc comment for this type?

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
Can you please update the version in version.go to v3.3.3?

**Comment by jonasrichard** [comment]
As a feature request - I am not a staticcheck expert - in case of missing documentation of public variables, consts, funcs and structs staticcheck should fail the build.

**Review by jhendrixMSFT — APPROVED** [review]

**Review by catalinaperalta — APPROVED** [review]

**Pull request merged**